repro/Proxy.cxx

```diff
@@ -462,8 +462,24 @@ Proxy::thread()
                }
```

Comment:
https://github.com/resiprocate/resiprocate/issues/209#issuecomment-1016596117

> please also consider that there are some TFM test cases for TimerC that may need adjusting as well.

@sgodin - none of the existing TimerC tests seem to be failing, at least I got a PASS on the `tfm/repro/sanityTests.cxx` test suite.

I have tried to create a new TimerC test case with serial forking by using `testSequentialQValueInvite()` as a starting point, but I realised that it won't be as easy due to conflict between TimerC logic and `QValueCancelBetweenForkGroups`/`QValueMsBeforeCancel` configuration options. I couldn't find a way to tweak Repro configuration for a single test case so I abandoned the idea of the adding a new test case.

Reply:
Right, I see what you mean.  QValueTargetHandler is created once and once the settings are passed in they are not changeable.  These are the options I can see:
1.  See if any tests rely on the QValueMsBeforeCancel setting being 5000 (from TestRepro.cxx line 49), if not then change it to what you need.
2. Add methods to QValueTargetHandler to allow the settings to be changed at runtime.
3. Add a way to restart the "proxy" under test, then in the test case change the value to what you need, restart the proxy, run the test, change back and restart the proxy again. 
4. Skip adding a unit test for this new behavior.

Options 1 or 2 seem best.

Reply:
1. I've increased `QValueMsBeforeCancel` value from 5 seconds to 10 seconds and got 7 test failures (oddly enough, failures were only reflected in the `sanityTest.log` file, while testsuite summary in the console was suggesting all tests have passed). I could fix timeouts in the 7 affected tests and set TimerC value to be less than 10 seconds, but then it would change the test behaviour - client transactions would be cancelled sooner and due to TimerC expiry and  not because of the built-in QValueTargetHandler logic. This makes me think that a dynamic way to control Repro config is needed to maintain existing test logic separately from the new TimerC test logic.
2. Not sure how this can be done without exposing access to processor chains outside the `Proxy` class (which doesn't seem like a good idea on it's own).
3. Could probably be achieved but to ensure that one test failure doesn't cause a knock-on effect, it will probably be best to perform configuration reset to default values in the `ReproFixture::tearDown()`, which may add significant overhead to the overall test suite run time (due to Repro restart after each test case).

@sgodin not seeing any clear winners here (if you don't count option 4). Any thoughts on the above analysis? Am I missing something or are you willing to accept some of the above-mentioned trade-offs?

Reply:
1.  Seems odd you get failures that only show up in the logs and not in the summary.  Do those failures go away when you put the QValueMsBeforeCancel back to 5 seconds?  Is it possible to write a unit test for your new timerC logic that only dispatches one set of targets (ie: parallel forking), and not have QValueMsBeforeCancel get used.  I am assuming if there is no 2nd group of targets to dispatch then the it won't cancel, but I'm not sure offhand.
2. It should be pretty easy to just keep a pointer around to the QValueTargetHandler when it is added (TestRepro.cxx line 107/108), then call a method on it to change the settings when needed.
3. I don't like this suggestion anyway, let's forget it.  :)
4.  I'd like to ensure 1 or 2 aren't easy solutions before we go with this.